### PR TITLE
CEDS-2091 Endpoint to handle ILE query

### PR DESCRIPTION
**Pull request by PatrykRudnicki**

### CODE REVIEW

File: app/uk/gov/hmrc/exports/movements/services/ILEMapper.scala [hunk 1]
```diff
@@ -101,4 +101,14 @@ class ILEMapper @Inject()(clock: Clock) {
     case ASSOCIATE_DUCR | DISASSOCIATE_DUCR => "D"
     case ASSOCIATE_MUCR | DISASSOCIATE_MUCR => "M"
   }
+
+  def generateIleQuery(ucrBlock: uk.gov.hmrc.exports.movements.models.notifications.UcrBlock): NodeSeq =
```
Comment (PatrykRudnicki) [inline app/uk/gov/hmrc/exports/movements/services/ILEMapper.scala]: Sorry for whole import here, but we have one `UcrBlock` in scope from `wco-dec`

File: app/uk/gov/hmrc/exports/movements/models/movements/IleQueryRequest.scala [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+ */
+
+package uk.gov.hmrc.exports.movements.models.movements
+
+import play.api.libs.json.Json
+import uk.gov.hmrc.exports.movements.controllers.util.JSONResponses
+import uk.gov.hmrc.exports.movements.models.UserIdentification
+import uk.gov.hmrc.exports.movements.models.notifications.UcrBlock
+
+case class IleQueryRequest(override val eori: String, override val providerId: Option[String], ucrBlock: UcrBlock) extends UserIdentification
```
Comment (MaciejRewera) [inline app/uk/gov/hmrc/exports/movements/models/movements/IleQueryRequest.scala]: 👍

File: test/unit/uk/gov/hmrc/exports/movements/base/UnitTestMockBuilder.scala [hunk 3]
```diff
@@ -37,7 +37,7 @@ import scala.xml.NodeSeq
 object UnitTestMockBuilder extends MockitoSugar {
 
   val dummyWriteResultSuccess: WriteResult =
-    DefaultWriteResult(ok = true, n = 1, writeErrors = Seq.empty, writeConcernError = None, code = None, errmsg = None)
+    LastError(true, None, None, None, 0, None, false, None, None, false, None, None)
```
Comment (MaciejRewera) [inline test/unit/uk/gov/hmrc/exports/movements/base/UnitTestMockBuilder.scala]: 👍

File: test/unit/uk/gov/hmrc/exports/movements/controllers/IleQueryControllerSpec.scala [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+import scala.concurrent.ExecutionContext.global
+import scala.concurrent.Future
+
+class IleQueryControllerSpec extends WordSpec with MustMatchers with MockitoSugar with ScalaFutures with BeforeAndAfterEach {
+
+  private val ileQueryService = mock[IleQueryService]
+
+  private val controller = new IleQueryController(ileQueryService, stubControllerComponents())(global)
+
+  override protected def afterEach(): Unit = {
```
Comment (MaciejRewera) [inline test/unit/uk/gov/hmrc/exports/movements/controllers/IleQueryControllerSpec.scala]: My advise would be to `reset` mocks before each test as well. This way you make sure you have your mocks in a state you expect.

Comment (PatrykRudnicki) [reply test/unit/uk/gov/hmrc/exports/movements/controllers/IleQueryControllerSpec.scala]: I'm creating mocks above and not reusing them, so I think `afterEach` is enough

File: test/unit/uk/gov/hmrc/exports/movements/services/IleQueryServiceSpec.scala [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+
+  val ileQueryService = new IleQueryService(ileMapper, submissionRepository, ileConnector)(global)
+
+  val conversationId = "conversationId"
+  val ileQueryRequest = IleQueryRequest("GB1234564654", Some("12345"), UcrBlock("9GB025115188654-IAZ1", "D"))
+
+  val writeResult = LastError(true, None, None, None, 0, None, false, None, None, false, None, None)
+
+  override protected def afterEach(): Unit = {
+    reset(submissionRepository, ileConnector)
```
Comment (MaciejRewera) [inline test/unit/uk/gov/hmrc/exports/movements/services/IleQueryServiceSpec.scala]: Same thing here with resetting.

File: test/unit/uk/gov/hmrc/exports/movements/services/IleQueryServiceSpec.scala [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+      when(ileConnector.submit(any(), any())(any()))
+        .thenReturn(Future.successful(CustomsInventoryLinkingResponse(ACCEPTED, Some(conversationId))))
+      when(submissionRepository.insert(any())(any())).thenReturn(Future.successful(writeResult))
+
+      val result = ileQueryService.submit(ileQueryRequest)(hc).futureValue
+
+      result mustBe conversationId
+    }
+
+    "fail when non accepted response from ILE" in {
```
Comment (MaciejRewera) [inline test/unit/uk/gov/hmrc/exports/movements/services/IleQueryServiceSpec.scala]: What about a case when writing to the DB did not succeed?

File: test/unit/uk/gov/hmrc/exports/movements/services/IleQueryServiceSpec.scala [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+  val ileMapper = new ILEMapper(Clock.fixed(Instant.now(), ZoneOffset.UTC))
+  val submissionRepository = mock[SubmissionRepository]
+  val ileConnector = mock[CustomsInventoryLinkingExportsConnector]
+
+  val ileQueryService = new IleQueryService(ileMapper, submissionRepository, ileConnector)(global)
+
+  val conversationId = "conversationId"
+  val ileQueryRequest = IleQueryRequest("GB1234564654", Some("12345"), UcrBlock("9GB025115188654-IAZ1", "D"))
+
+  val writeResult = LastError(true, None, None, None, 0, None, false, None, None, false, None, None)
```
Comment (MaciejRewera) [inline test/unit/uk/gov/hmrc/exports/movements/services/IleQueryServiceSpec.scala]: I think you've defined `dummyWriteResultSuccess` earlier. Please use it.

File: test/unit/uk/gov/hmrc/exports/movements/services/IleQueryServiceSpec.scala [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+  val hc = mock[HeaderCarrier]
+
+  val ileMapper = new ILEMapper(Clock.fixed(Instant.now(), ZoneOffset.UTC))
+  val submissionRepository = mock[SubmissionRepository]
+  val ileConnector = mock[CustomsInventoryLinkingExportsConnector]
+
+  val ileQueryService = new IleQueryService(ileMapper, submissionRepository, ileConnector)(global)
+
+  val conversationId = "conversationId"
+  val ileQueryRequest = IleQueryRequest("GB1234564654", Some("12345"), UcrBlock("9GB025115188654-IAZ1", "D"))
```
Comment (MaciejRewera) [inline test/unit/uk/gov/hmrc/exports/movements/services/IleQueryServiceSpec.scala]: Is there a particular reason why you are not using values from `CommonTestData` ?

File: test/utils/testdata/ConsolidationTestData.scala [hunk 9]
```diff
@@ -75,6 +75,14 @@ object ConsolidationTestData {
...
     </inventoryLinkingConsolidationRequest>
 
+  val exampleIleQueryRequestXml: Elem =
+    <inventoryLinkingQueryRequest>
+      <queryUCR>
+        <ucr>{ucr}</ucr>
+        <ucrType>D</ucrType>
+      </queryUCR>
+    </inventoryLinkingQueryRequest>
+
```
Comment (MaciejRewera) [inline test/utils/testdata/ConsolidationTestData.scala]: 👍

File: test/unit/uk/gov/hmrc/exports/movements/controllers/IleQueryControllerSpec.scala (hunk 4 above)
Comment (PatrykRudnicki) [reply test/unit/uk/gov/hmrc/exports/movements/controllers/IleQueryControllerSpec.scala]: Fixed

File: test/unit/uk/gov/hmrc/exports/movements/services/IleQueryServiceSpec.scala (hunk 5 above)
Comment (PatrykRudnicki) [reply test/unit/uk/gov/hmrc/exports/movements/services/IleQueryServiceSpec.scala]: Fixed

File: test/unit/uk/gov/hmrc/exports/movements/services/IleQueryServiceSpec.scala (hunk 7 above)
Comment (PatrykRudnicki) [reply test/unit/uk/gov/hmrc/exports/movements/services/IleQueryServiceSpec.scala]: Fixed

File: test/unit/uk/gov/hmrc/exports/movements/services/IleQueryServiceSpec.scala (hunk 8 above)
Comment (PatrykRudnicki) [reply test/unit/uk/gov/hmrc/exports/movements/services/IleQueryServiceSpec.scala]: I didn't have it in scope, can change it

Comment (PatrykRudnicki) [reply test/unit/uk/gov/hmrc/exports/movements/services/IleQueryServiceSpec.scala]: Fixed

File: test/unit/uk/gov/hmrc/exports/movements/services/IleQueryServiceSpec.scala (hunk 6 above)
Comment (PatrykRudnicki) [reply test/unit/uk/gov/hmrc/exports/movements/services/IleQueryServiceSpec.scala]: We don't have tests for that case. Moreover we're not handling this situation. Not sure what is the correct output of failure like that. It'll be handled by general error

**Comment by PatrykRudnicki** [comment]
@MaciejRewera Fixed all comments except DB failure

**Pull request merged**